File: android/core/calc/src/main/kotlin/com/stocktracker/core/calc/RowDerivation.kt
package com.stocktracker.core.calc

import com.stocktracker.core.model.DivTaxOverrides
import com.stocktracker.core.model.DividendEvent
import com.stocktracker.core.model.FX_CONVERTED_TICKERS
import com.stocktracker.core.model.ManualPriceEntry
import com.stocktracker.core.model.Position
import com.stocktracker.core.model.PortfolioRow
import com.stocktracker.core.model.Quote

/** CZK-base FX conversion, injected so this module stays pure and testable. */
typealias Convert = (Double, String, String) -> Double

private fun isOpenLot(lot: Position): Boolean {
    val sellPrice = lot.sellPrice
    val sellDate = lot.sellDate
    return sellPrice == null || sellPrice == 0.0 || sellDate == null || sellDate.isEmpty()
}

private fun isClosedLot(lot: Position): Boolean =
    lot.sellPrice != null && !lot.sellDate.isNullOrEmpty()

/**
 * One aggregated ticker row. Mirrors the `rows` useMemo in
 * src/components/PortfolioContent.tsx line-for-line — see that file before
 * changing anything here, this is a faithful port, not a redesign.
 *
 * `quote`/`loading`/`error`/`manual` are the *raw* per-ticker values (before
 * the source's isClosed-nulling) — this function applies that nulling
 * internally, exactly like the source does by shadowing `quote` etc. once
 * at the top of its row-building closure.
 */
fun deriveRow(
    lots: List<Position>,
    quoteRaw: Quote?,
    loadingRaw: Boolean,
    errorRaw: String?,
    manualRaw: ManualPriceEntry?,
    dividends: List<DividendEvent>,
    taxOverrides: DivTaxOverrides = emptyMap(),
    today: String,
    convert: Convert,
): PortfolioRow {
    require(lots.isNotEmpty()) { "lots must not be empty" }
    val ticker = lots.first().ticker

    val openLots = lots.filter(::isOpenLot)
    val closedLots = lots.filter(::isClosedLot)
    val isClosed = openLots.isEmpty()

    val rowCurrency = lots.first().currency
    fun toRow(amount: Double, lotCurrency: String): Double = convert(amount, lotCurrency, rowCurrency)

    val totalQty = lots.sumOf { it.quantity }
    val openQty = openLots.sumOf { it.quantity }
    val totalCost = lots.sumOf { toRow(it.buyPrice * it.quantity, it.currency) }
    val openCost = openLots.sumOf { toRow(it.buyPrice * it.quantity, it.currency) }
    val avgBuyPrice = totalCost / totalQty
    val firstBuyDate = lots.minOf { it.buyDate }

    // Source nulls these out once isClosed, then uses that nulled value everywhere below.
    val quote = if (isClosed) null else quoteRaw
    val isLoading = !isClosed && loadingRaw
    val error = if (isClosed) null else errorRaw
    val manual = if (isClosed) null else manualRaw
    val priceIsManual = !isClosed && quote == null && manual != null

    val fxEntry = FX_CONVERTED_TICKERS[ticker.uppercase()]
    val nativeCurrency = fxEntry?.fxTicker?.take(3) ?: quote?.currency ?: rowCurrency

    val closedQty = closedLots.sumOf { it.quantity }
    val avgSellPrice = if (closedLots.isNotEmpty()) {
        closedLots.sumOf { toRow((it.sellPrice ?: 0.0) * it.quantity, it.currency) } / closedQty
    } else 0.0
    val openAvgBuy = if (openQty > 0) openCost / openQty else 0.0
    val quotePrice = quote?.let { toRow(it.price, it.currency) }
    val currentPrice = if (isClosed) avgSellPrice else (quotePrice ?: manual?.price ?: avgBuyPrice)
    val currentValue = if (isClosed) 0.0 else currentPrice * openQty

    val divCurrency = dividends.firstOrNull()?.currency ?: rowCurrency
    val dividendIncome = toRow(calcNetDividends(lots, dividends, ticker, taxOverrides), divCurrency)

    val realizedPnl = closedLots.sumOf { toRow(((it.sellPrice ?: 0.0) - it.buyPrice) * it.quantity, it.currency) }
    val unrealizedPnl = if (isClosed) 0.0 else (currentPrice - openAvgBuy) * openQty
    val pricePnl = realizedPnl + unrealizedPnl
    val totalReturn = pricePnl + dividendIncome
    val pnlPercent = if (totalCost > 0) pricePnl / totalCost * 100 else 0.0

    val hasUsablePrice = isClosed || (!isLoading && (quote != null || manual != null))
    val irr: Double? = if (!hasUsablePrice) null else {
        val flows = mutableListOf<CashFlow>()
        lots.forEach { flows.add(CashFlow(it.buyDate, -toRow(it.buyPrice * it.quantity, it.currency))) }
        closedLots.forEach { flows.add(CashFlow(it.sellDate!!, toRow(it.sellPrice!! * it.quantity, it.currency))) }
        dividends.forEach { div ->
            val shares = lots.filter { lot ->
                val sd = lot.sellDate
                lot.buyDate <= div.date && (sd == null || sd > div.date)
            }.sumOf { it.quantity }
            if (shares != 0.0) {
                val rate = taxOverrides["${ticker.uppercase()}::${div.date}"] ?: getDividendTaxRate(ticker)
                flows.add(CashFlow(div.date, toRow(shares * div.amount * (1 - rate), div.currency)))
            }
        }
        if (!isClosed) flows.add(CashFlow(today, currentValue))
        xirr(flows)
    }

    val dailyChange = if (isClosed || quote == null) 0.0 else toRow(quote.change, quote.currency) * openQty

    return PortfolioRow(
        ids = lots.map { it.id },
        ticker = ticker,
        name = lots.first().name,
        type = lots.first().type,
        currency = rowCurrency,
        nativeCurrency = nativeCurrency,
        lots = lots.size,
        positions = lots.sortedBy { it.buyDate },
        totalQuantity = totalQty,
        avgBuyPrice = avgBuyPrice,
        firstBuyDate = firstBuyDate,
        currentPrice = currentPrice,
        currentValue = currentValue,
        costBasis = totalCost,
        pnl = pricePnl,
        pnlPercent = pnlPercent,
        dividendIncome = dividendIncome,
        totalReturn = totalReturn,
        loading = isLoading,
        error = error,
        priceIsManual = priceIsManual,
        manualPriceDate = manual?.updatedAt,
        irr = irr,
        isClosed = isClosed,
        dailyChange = dailyChange,
    )
}

/**
 * Portfolio-wide IRR — a separate aggregation over every individual position,
 * not an average of per-row IRRs (XIRR isn't additive). Mirrors the
 * `portfolioIrr` useMemo in src/components/PortfolioContent.tsx.
 */
fun computePortfolioIrr(
    positions: List<Position>,
    rows: List<PortfolioRow>,
    dividendsByTicker: Map<String, List<DividendEvent>>,
    taxOverrides: DivTaxOverrides = emptyMap(),
    displayCurrency: String,
    today: String,
    convert: Convert,
): Double? {
    if (positions.isEmpty()) return null

    val anyLoading = rows.any { it.loading }
    val anyMissingPrice = rows.any { it.error == null && it.irr == null && !it.loading }
    if (anyLoading || anyMissingPrice) return null

    fun toDc(amount: Double, currency: String) = convert(amount, currency, displayCurrency)

    val totalCurrentValue = rows.sumOf { toDc(it.currentValue, it.currency) }

    val flows = mutableListOf<CashFlow>()
    positions.forEach { pos -> flows.add(CashFlow(pos.buyDate, -toDc(pos.buyPrice * pos.quantity, pos.currency))) }
    positions.filter(::isClosedLot).forEach { pos ->
        flows.add(CashFlow(pos.sellDate!!, toDc(pos.sellPrice!! * pos.quantity, pos.currency)))
    }
    positions.forEach { pos ->
        val divs = dividendsByTicker[pos.ticker.uppercase()] ?: emptyList()
        divs.forEach { div ->
            val sellDate = pos.sellDate
            if (pos.buyDate <= div.date && (sellDate == null || sellDate > div.date)) {
                val rate = taxOverrides["${pos.ticker.uppercase()}::${div.date}"] ?: getDividendTaxRate(pos.ticker)
                flows.add(CashFlow(div.date, toDc(pos.quantity * div.amount * (1 - rate), div.currency)))
            }
        }
    }
    flows.add(CashFlow(today, totalCurrentValue))

    return xirr(flows)
}

File: src/utils/rowDerivation.ts
import type { PortfolioRow, Position, Quote } from '../types'
import type { DividendEvent } from './dividends'
import { xirr } from './xirr'
import { calcNetDividends, getDividendTaxRate } from './dividends'
import { FX_CONVERTED_TICKERS } from '../data/fxConvertedTickers'
import { dailyChange as dailyChangeOf } from './dailyChange'

export type Convert = (amount: number, from: string, to: string) => number

/** A lot is sold once it has both a sell date and a sell price — a price of 0 (a worthless
 *  delisting) still counts as sold. Mirrors `isClosedLot` in core/calc/Lots.kt. */
export const isClosedLot = (l: Position): boolean => l.sellPrice != null && !!l.sellDate
export const isOpenLot = (l: Position): boolean => !isClosedLot(l)

export interface DeriveRowInput {
  lots: Position[]
  quote?: Quote
  loading: boolean
  error: string | null
  manual?: { price: number; updatedAt?: string }
  dividends: DividendEvent[]
  taxOverrides: Record<string, number>
  /** YYYY-MM-DD — injected so the IRR terminal flow is deterministic under test. */
  today: string
  convert: Convert
  /** Midnight anchor for this ticker, in the quote's own currency. */
  anchor?: { price: number | null; lastTradedAt: number | null }
  /** Quote currency -> display currency at local midnight. */
  anchorFx?: number | null
  displayCurrency: string
}

/**
 * One aggregated ticker row. Mirrored parameter-for-parameter by `deriveRow` in
 * android/core/calc/.../RowDerivation.kt; test-fixtures/parity pins the two together.
 * The raw quote/loading/error/manual are nulled here once the row is fully closed.
 */
export function deriveRow(i: DeriveRowInput): PortfolioRow {
  const { lots, dividends: tickerDivs, taxOverrides, convert, displayCurrency } = i
  const ticker = lots[0].ticker

  const openLots = lots.filter(isOpenLot)
  const closedLots = lots.filter(isClosedLot)
  const isClosed = openLots.length === 0

  const rowCurrency = lots[0].currency
  // ponytail: converts lot amount to row currency so mixed-currency lots aggregate correctly
  const toRow = (amount: number, lotCurrency: string) => convert(amount, lotCurrency, rowCurrency)

  const totalQty = lots.reduce((s, p) => s + p.quantity, 0)
  const openQty = openLots.reduce((s, p) => s + p.quantity, 0)
  const totalCost = lots.reduce((s, p) => s + toRow(p.buyPrice * p.quantity, p.currency), 0)
  const openCost = openLots.reduce((s, p) => s + toRow(p.buyPrice * p.quantity, p.currency), 0)
  const avgBuyPrice = totalQty > 0 ? totalCost / totalQty : 0
  const firstBuyDate = [...lots].sort((a, b) => a.buyDate.localeCompare(b.buyDate))[0].buyDate

  const quote = isClosed ? undefined : i.quote
  const isLoading = !isClosed && i.loading
  const error = isClosed ? null : i.error
  const manual = isClosed ? undefined : i.manual
  const priceIsManual = !isClosed && !quote && !!manual

  // The currency the instrument itself trades in — independent of what
  // currency a given lot's buy price happened to be recorded in (e.g. a
  // CZK brokerage statement for a EUR-denominated ETC like 4GLD.DE).
  // FX-converted tickers (fxConvertedTickers.ts) fetch their price in a
  // foreign currency and multiply by an FX pair to get a CZK value, so
  // their true native currency is the fxTicker's base (e.g. EURCZK=X → EUR).
  const fxEntry = FX_CONVERTED_TICKERS[ticker.toUpperCase()]
  const nativeCurrency = fxEntry ? fxEntry.fxTicker.slice(0, 3) : (quote?.currency ?? rowCurrency)

  const avgSellPrice = closedLots.length > 0
    ? closedLots.reduce((s, l) => s + toRow(l.sellPrice! * l.quantity, l.currency), 0) / closedLots.reduce((s, l) => s + l.quantity, 0)
    : 0
  const openAvgBuy = openQty > 0 ? openCost / openQty : 0
  // Quotes arrive in the asset's native currency (JPY for .T, EUR for .AS, …),
  // which can differ from the lot currency the broker statement was priced in.
  const quotePrice = quote ? toRow(quote.price, quote.currency) : undefined
  const currentPrice = isClosed
    ? avgSellPrice
    : (quotePrice ?? manual?.price ?? avgBuyPrice)
  const currentValue = isClosed ? 0 : currentPrice * openQty

  // Dividend amounts are per-share in the ticker's native currency (Yahoo meta.currency)
  const divCurrency = tickerDivs[0]?.currency ?? rowCurrency
  const dividendIncome = toRow(calcNetDividends(lots, tickerDivs, ticker, taxOverrides), divCurrency)

  const realizedPnl = closedLots.reduce((s, l) => s + toRow((l.sellPrice! - l.buyPrice) * l.quantity, l.currency), 0)
  const unrealizedPnl = isClosed ? 0 : (currentPrice - openAvgBuy) * openQty
  const pricePnl = realizedPnl + unrealizedPnl
  const totalReturn = pricePnl + dividendIncome

  const hasUsablePrice = isClosed || (!isLoading && (!!quote || !!manual))
  const irrValue = hasUsablePrice
    ? xirr([
        ...lots.map((p) => ({ date: new Date(p.buyDate), amount: -toRow(p.buyPrice * p.quantity, p.currency) })),
        ...closedLots.map((l) => ({ date: new Date(l.sellDate!), amount: toRow(l.sellPrice! * l.quantity, l.currency) })),
        ...tickerDivs.flatMap((div) => {
          const shares = lots
            .filter((l) => l.buyDate <= div.date && (!l.sellDate || l.sellDate > div.date))
            .reduce((s, l) => s + l.quantity, 0)
          if (shares === 0) return []
          const rate = taxOverrides[`${ticker.toUpperCase()}::${div.date}`] ?? getDividendTaxRate(ticker)
          return [{ date: new Date(div.date), amount: toRow(shares * div.amount * (1 - rate), div.currency ?? rowCurrency) }]
        }),
        ...(isClosed ? [] : [{ date: new Date(i.today), amount: currentValue }]),
      ])
    : null

  const dailyChange = isClosed || !quote ? 0 : toRow(quote.change, quote.currency) * openQty

  // Today's change, measured from midnight in the user's own zone rather than from the
  // exchange's previous close — see docs/superpowers/specs/2026-09-19-timezone-anchored-
  // daily-change-design.md. Worked in the quote's own currency: the anchor price and the
  // FX anchor are both quoted there, and converting either one through the row currency
  // first would mix a spot rate into a figure that is meant to be anchored.
  const quoteCurrency = quote?.currency ?? rowCurrency
  const anchor = isClosed || !quote ? undefined : i.anchor
  const anchorFx = isClosed || !quote ? undefined : i.anchorFx
  const daily = dailyChangeOf({
    quantity: isClosed || !quote ? 0 : openQty,
    currentPrice: quote?.price ?? 0,
    anchorPrice: anchor?.price ?? null,
    currentFx: convert(1, quoteCurrency, displayCurrency),
    anchorFx: anchorFx ?? null,
    // Reconstructed from the figure the quote already carries, so the fallback keeps
    // reporting exactly what the app showed before this feature existed.
    prevClose: quote ? quote.price - quote.change : null,
  })

  return {
    ids: lots.map((p) => p.id),
    ticker,
    name: lots[0].name,
    type: lots[0].type,
    currency: lots[0].currency,
    nativeCurrency,
    lots: lots.length,
    positions: [...lots].sort((a, b) => a.buyDate.localeCompare(b.buyDate)),
    totalQuantity: totalQty,
    openQuantity: openQty,
    avgBuyPrice,
    firstBuyDate,
    currentPrice,
    currentValue,
    costBasis: totalCost,
    pnl: pricePnl,
    pnlPercent: totalCost > 0 ? (pricePnl / totalCost) * 100 : 0,
    dividendIncome,
    totalReturn,
    loading: isLoading,
    error,
    priceIsManual,
    manualPriceDate: manual?.updatedAt,
    irr: irrValue,
    isClosed,
    dailyChange,
    dailyChangeDisplay: daily.change,
    dailyChangePercent: daily.changePercent,
    dailyPriceOnlyDisplay: daily.priceOnlyChange,
    dailyChangeMethod: daily.method,
    lastTradedAt: anchor?.lastTradedAt ?? null,
  }
}

/** Groups lots by ticker in first-seen order — the row order both apps start from. */
export function groupByTicker(positions: Position[]): Position[][] {
  const groups = new Map<string, Position[]>()
  positions.forEach((p) => groups.set(p.ticker, [...(groups.get(p.ticker) ?? []), p]))
  return [...groups.values()]
}

/**
 * Portfolio-wide IRR — one XIRR over every lot, not an average of row IRRs. Mirrors
 * `computePortfolioIrr` in android/core/calc/.../RowDerivation.kt.
 */
export function computePortfolioIrr(
  positions: Position[],
  rows: PortfolioRow[],
  dividends: Map<string, DividendEvent[]>,
  taxOverrides: Record<string, number>,
  displayCurrency: string,
  today: string,
  convert: Convert,
): number | null {
  if (positions.length === 0) return null
  const anyLoading = rows.some((r) => r.loading)
  // Waits for open rows still missing a price. A closed row can legitimately have no IRR
  // (a total loss has no root), and must not blank the whole portfolio's figure.
  const anyMissingPrice = rows.some((r) => !r.isClosed && !r.error && r.irr === null && !r.loading)
  if (anyLoading || anyMissingPrice) return null

  const toDC = (amount: number, currency: string) => convert(amount, currency, displayCurrency)
  const totalCurrentValue = rows.reduce((s, r) => s + toDC(r.currentValue, r.currency), 0)

  const divCashFlows: { date: Date; amount: number }[] = []
  positions.forEach((pos) => {
    const divs = dividends.get(pos.ticker.toUpperCase()) ?? []
    divs.forEach((div) => {
      if (pos.buyDate <= div.date && (!pos.sellDate || pos.sellDate > div.date)) {
        const rate = taxOverrides[`${pos.ticker.toUpperCase()}::${div.date}`] ?? getDividendTaxRate(pos.ticker)
        divCashFlows.push({ date: new Date(div.date), amount: toDC(pos.quantity * div.amount * (1 - rate), div.currency ?? pos.currency) })
      }
    })
  })

  const sellCashFlows = positions
    .filter((p) => p.sellPrice != null && p.sellDate)
    .map((p) => ({ date: new Date(p.sellDate!), amount: toDC(p.sellPrice! * p.quantity, p.currency) }))

  return xirr([
    ...positions.map((p) => ({ date: new Date(p.buyDate), amount: -toDC(p.buyPrice * p.quantity, p.currency) })),
    ...sellCashFlows,
    ...divCashFlows,
    { date: new Date(today), amount: totalCurrentValue },
  ])
}
